File: php/example_code/elastictranscoder/HlsJobCreationSample.php
<?php
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

/*
 *  ABOUT THIS PHP SAMPLE: This sample is part of the Elastic Transcoder Developer Guide topic at
 *  https://docs.aws.amazon.com/elastictranscoder/latest/developerguide/introduction.html
 *
 */
// snippet-start:[elastictranscoder.php.creation.sample]
// Path to your PHP autoload.  If you are using a phar installation, this is the
// path to your aws.phar file.
require_once 'path/to/autoload.php';

use Aws\ElasticTranscoder\ElasticTranscoderClient;

// Region where the sample will be run.
$region = 'us-east-1';

// HLS Presets that will be used to create an adaptive bitrate playlist.
$hls_64k_audio_preset_id = '1351620000001-200071';
$hls_0400k_preset_id     = '1351620000001-200050';
$hls_0600k_preset_id     = '1351620000001-200040';
$hls_1000k_preset_id     = '1351620000001-200030';
$hls_1500k_preset_id     = '1351620000001-200020';
$hls_2000k_preset_id     = '1351620000001-200010';

$hls_presets = array(
  'hlsAudio' => $hls_64k_audio_preset_id,
  'hls0400k' => $hls_0400k_preset_id,
  'hls0600k' => $hls_0600k_preset_id,
  'hls1000k' => $hls_1000k_preset_id,
  'hls1500k' => $hls_1500k_preset_id,
  'hls2000k' => $hls_2000k_preset_id,
);

// HLS Segment duration that will be targeted.
$segment_duration = '2';

//All outputs will have this prefix prepended to their output key.
$output_key_prefix = 'elastic-transcoder-samples/output/hls/';

// Create the client for Elastic Transcoder.
$transcoder_client = ElasticTranscoderClient::factory(['region' => $region, 'default_caching_config' => '/tmp']);

function create_hls_job(
    $transcoder_client,
    $pipeline_id,
    $input_key,
    $output_key_prefix,
    $hls_presets,
    $segment_duration
) {
  // Set up the job input using the provided input key.
    $input = ['Key' => $input_key];

  //Set up the job outputs using the HLS presets.
    $output_key = hash('sha256', utf8_encode($input_key));

  // Specify the outputs based on the hls presets array specified.
    $outputs = [];
    foreach ($hls_presets as $prefix => $preset_id) {
        $outputs[] = ['Key' => "$prefix/$output_key", 'PresetId' => $preset_id, 'SegmentDuration' => $segment_duration];
    }

  // Setup master playlist which can be used to play using adaptive bitrate.
    $playlist = [
        'Name' => 'hls_' . $output_key,
        'Format' => 'HLSv3',
        'OutputKeys' => array_map(function ($x) {
            return $x['Key'];
        }, $outputs)
    ];

  // Create the job.
    $create_job_request = [
        'PipelineId' => $pipeline_id,
        'Input' => $input,
        'Outputs' => $outputs,
        'OutputKeyPrefix' => "$output_key_prefix$output_key/",
        'Playlists' => [$playlist]
    ];
    $create_job_result = $transcoder_client->createJob($create_job_request)->toArray();
    return $job = $create_job_result['Job'];
}

if ($_SERVER['REQUEST_METHOD'] == 'GET') {
  // If the request method is GET, return the form which will allow the user to
  // specify pipeline and input key.
    echo "Create an Elastic Transcoder HLS job.<br><form action=\"HlsJobCreationSample.php\" ";
    echo "method=\"POST\">Pipeline Id: <input name=\"pipelineid\" type=\"text\"/> (<a href=\"";
    echo "http://docs.aws.amazon.com/elastictranscoder/latest/developerguide/sample-code.html#php-pipeline\">";
    echo "Create an Elastic Transcoder Pipeline</a>)<br>Input key: <input name=\"inputkey\" type=\"text\" /><br>";
    echo "<input type=\"submit\" value=\"Create Job\" /></form>";
} elseif ($_SERVER['REQUEST_METHOD'] == 'POST') {
  // If the request method is POST, create an HLS job using the posted data.
    $job = create_hls_job(
        $transcoder_client,
        $_POST['pipelineid'],
        $_POST['inputkey'],
        $output_key_prefix,
        $hls_presets,
        $segment_duration
    );

  // Output the result.
    echo '<PRE>';
    echo "HLS job has been created:\n";
    echo json_encode($job, JSON_PRETTY_PRINT);
    echo '</PRE>';
}

// snippet-end:[elastictranscoder.php.creation.sample]

File: java/example_code/elastictranscoder/src/main/java/aws/example/elastictranscoder/HlsJobCreationSample.java
//snippet-sourcedescription:[HlsJobCreationSample.java demonstrates how to create an HLS job.]
//snippet-keyword:[Java]
//snippet-sourcesyntax:[java]
//snippet-keyword:[Code Sample]
//snippet-keyword:[Amazon Elastic Transcoder]
//snippet-service:[elastictranscoder]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[]
//snippet-sourceauthor:[AWS]
/*
 * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
// snippet-start:[elastictranscoder.java.create_hls_job.import]
package com.amazonaws.services.elastictranscoder.samples;

import java.util.Arrays;
import java.util.List;

import com.amazonaws.services.elastictranscoder.AmazonElasticTranscoder;
import com.amazonaws.services.elastictranscoder.AmazonElasticTranscoderClient;
import com.amazonaws.services.elastictranscoder.model.CreateJobOutput;
import com.amazonaws.services.elastictranscoder.model.CreateJobPlaylist;
import com.amazonaws.services.elastictranscoder.model.CreateJobRequest;
import com.amazonaws.services.elastictranscoder.model.Job;
import com.amazonaws.services.elastictranscoder.model.JobInput;
import com.amazonaws.services.elastictranscoder.samples.utils.TranscoderSampleUtilities;

/**
 * The purpose of this sample is to show how to create an HLS job and HLS
 * playlist file which can be used to play an adaptive bitrate stream.  HLS
 * playlists can be played in the browser using tools such as jwplayer or using
 * media players such as VLC.
 */
public class HlsJobCreationSample {

    // This is the ID of the Elastic Transcoder pipeline that was created when
    // setting up your AWS environment:
    // http://docs.aws.amazon.com/elastictranscoder/latest/developerguide/sample-code.html#java-pipeline
    private static final String PIPELINE_ID = "Enter your pipeline id here.";
    
    // This is the name of the input key that you would like to transcode.
    private static final String INPUT_KEY = "Enter your input key here.";
    
    // HLS Presets that will be used to create an adaptive bitrate playlist.
    private static final String HLS_64K_AUDIO_PRESET_ID = "1351620000001-200071";
    private static final String HLS_0400K_PRESET_ID     = "1351620000001-200050";
    private static final String HLS_0600K_PRESET_ID     = "1351620000001-200040";
    private static final String HLS_1000K_PRESET_ID     = "1351620000001-200030";
    private static final String HLS_1500K_PRESET_ID     = "1351620000001-200020";
    private static final String HLS_2000K_PRESET_ID     = "1351620000001-200010";
    
    // HLS Segment duration that will be targeted.
    private static final String SEGMENT_DURATION = "2";
    
    // All outputs will have this prefix prepended to their output key.
    private static final String OUTPUT_KEY_PREFIX = "elastic-transcoder-samples/output/hls/";
    
    // Clients are built using the default credentials provider chain.  This
    // will attempt to get your credentials in the following order:
    //      1. Environment variables (AWS_ACCESS_KEY and AWS_SECRET_KEY).
    //      2. Java system properties (AwsCredentials.properties).
    //      3. Instance profile credentials on EC2 instances.
    private static final AmazonElasticTranscoder amazonElasticTranscoder = new AmazonElasticTranscoderClient();
    
    public static void main(String[] args) throws Exception {
        
        Job job = createElasticTranscoderHlsJob();
        System.out.println("HLS job has been created: " + job);
    }
    
    /**
     * Creates a job which outputs an HLS playlist for adaptive bitrate playback.
     * @return Job that was created by Elastic Transcoder.
     * @throws Exception
     */
    private static Job createElasticTranscoderHlsJob() throws Exception {
        
        // Setup the job input using the provided input key.
        JobInput input = new JobInput()
            .withKey(INPUT_KEY);
        
        // Setup the job outputs using the HLS presets.
        String outputKey = TranscoderSampleUtilities.inputKeyToOutputKey(INPUT_KEY);
        CreateJobOutput hlsAudio = new CreateJobOutput()
            .withKey("hlsAudio/" + outputKey)
            .withPresetId(HLS_64K_AUDIO_PRESET_ID)
            .withSegmentDuration(SEGMENT_DURATION);
        CreateJobOutput hls0400k = new CreateJobOutput()
            .withKey("hls0400k/" + outputKey)
            .withPresetId(HLS_0400K_PRESET_ID)
            .withSegmentDuration(SEGMENT_DURATION);
        CreateJobOutput hls0600k = new CreateJobOutput()
            .withKey("hls0600k/" + outputKey)
            .withPresetId(HLS_0600K_PRESET_ID)
            .withSegmentDuration(SEGMENT_DURATION);
        CreateJobOutput hls1000k = new CreateJobOutput()
            .withKey("hls1000k/" + outputKey)
            .withPresetId(HLS_1000K_PRESET_ID)
            .withSegmentDuration(SEGMENT_DURATION);
        CreateJobOutput hls1500k = new CreateJobOutput()
            .withKey("hls1500k/" + outputKey)
            .withPresetId(HLS_1500K_PRESET_ID)
            .withSegmentDuration(SEGMENT_DURATION);
        CreateJobOutput hls2000k = new CreateJobOutput()
            .withKey("hls2000k/" + outputKey)
            .withPresetId(HLS_2000K_PRESET_ID)
            .withSegmentDuration(SEGMENT_DURATION);
        List<CreateJobOutput> outputs = Arrays.asList(hlsAudio, hls0400k, hls0600k, hls1000k, hls1500k, hls2000k);
        
        // Setup master playlist which can be used to play using adaptive bitrate.
        CreateJobPlaylist playlist = new CreateJobPlaylist()
            .withName("hls_" + outputKey)
            .withFormat("HLSv3")
            .withOutputKeys(hlsAudio.getKey(), hls0400k.getKey(), hls0600k.getKey(), hls1000k.getKey(), hls1500k.getKey(), hls2000k.getKey());
        
        // Create the job.
        CreateJobRequest createJobRequest = new CreateJobRequest()
            .withPipelineId(PIPELINE_ID)
            .withInput(input)
            .withOutputKeyPrefix(OUTPUT_KEY_PREFIX + outputKey + "/")
            .withOutputs(outputs)
            .withPlaylists(playlist);
        return amazonElasticTranscoder.createJob(createJobRequest).getJob();
    }
}
// snippet-end:[elastictranscoder.java.create_hls_job.import]
